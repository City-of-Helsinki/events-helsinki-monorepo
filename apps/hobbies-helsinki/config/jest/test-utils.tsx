import type { ApolloCache, InMemoryCache } from '@apollo/client/cache';
import type { MockedResponse } from '@apollo/client/testing';
import type { RenderResult } from '@testing-library/react';
import { act, fireEvent, render } from '@testing-library/react';
import type { NextRouter } from 'next/router';
import Router from 'next/router';
import type { ReactElement } from 'react';
import React from 'react';
import wait from 'waait';

import { hobbiesApolloClient } from '../../src/domain/clients/hobbiesApolloClient';
import TestProviders from './TestProviders';

type CustomRender = {
  (
    ui: React.ReactElement,
    options?: {
      mocks?: MockedResponse[];
      cache?: ApolloCache<Record<string, unknown>> | InMemoryCache;
      routes?: string | string[];
    }
  ): CustomRenderResult;
};

export type CustomRenderResult = RenderResult & { router: NextRouter };

export const arrowUpKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const enterKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 13, key: 'Enter' });

export const escKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = (): boolean =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });

const customRender: CustomRender = (
  ui: ReactElement,
  { mocks = [], cache = hobbiesApolloClient.cache, routes = [] } = {}
) => {
  if (routes) {
    if (!Array.isArray(routes)) {
      routes = [routes];
    }
    routes.forEach((route) => Router.push(route));
  }

  // set locales so that the routing tests for default locale worked
  Router.locales = ['fi', 'en', 'sv'];

  const renderResult = render(ui, {
    wrapper: ({ children }) => (
      <TestProviders mocks={mocks} router={Router} cache={cache}>
        {children}
      </TestProviders>
    ),
  });
  return {
    ...renderResult,
    router: Router,
  };
};

// eslint-disable-next-line testing-library/no-unnecessary-act
const actWait = (amount?: number): Promise<void> => act(() => wait(amount));

// eslint-disable-next-line import/export
export { actWait, customRender as render };

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';
export { render as defaultRender } from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
