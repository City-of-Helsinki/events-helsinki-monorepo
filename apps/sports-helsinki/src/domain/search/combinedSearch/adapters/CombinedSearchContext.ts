import { EventTypeId } from '@events-helsinki/components/types';
import type { Router } from 'next/router';
import React, { createContext } from 'react';
import {
  initialCombinedSearchFormValues,
  initialEventSearchAdapterValues,
  initialVenueSearchAdapterValues,
} from '../constants';
import type { CombinedSearchAdapterInput } from '../types';
import type CombinedSearchFormAdapter from './CombinedSearchFormAdapter';

export type CombinedSearchContextType = {
  formValues: CombinedSearchAdapterInput;
  searchVariables: ReturnType<CombinedSearchFormAdapter['getSearchVariables']>;
  setFormValues: (values: Partial<CombinedSearchAdapterInput>) => void;
  setFormValue: (
    field: keyof CombinedSearchAdapterInput,
    value: CombinedSearchAdapterInput[keyof CombinedSearchAdapterInput]
  ) => void;
  updateRouteToSearchPage: (options?: Parameters<Router['push']>[2]) => void;
  resetFormValues: () => void;
};

export const CombinedSearchContext = createContext<CombinedSearchContextType>({
  formValues: { ...initialCombinedSearchFormValues },
  searchVariables: {
    venue: initialVenueSearchAdapterValues,
    event: {
      ...initialEventSearchAdapterValues,
      eventType: EventTypeId.General,
    },
    course: {
      ...initialEventSearchAdapterValues,
      eventType: EventTypeId.Course,
    },
  },
  setFormValues() {
    throw new Error(
      'setFormValues to set multiple values at once, is still unimplemented, when it should be implemented!'
    );
  },
  setFormValue() {
    throw new Error(
      'setFormValue for a single field is still unimplemented, when it should be implemented!'
    );
  },
  updateRouteToSearchPage() {
    throw new Error(
      'updateRouteToSearchPage is still unimplemented, when it should be implemented!'
    );
  },
  resetFormValues() {
    throw new Error(
      'resetFormValues is still unimplemented, when it should be implemented!'
    );
  },
});

export function useCombinedSearchContext() {
  const context = React.useContext(CombinedSearchContext);
  if (!context) {
    throw new Error(
      `Combined search components cannot be rendered outside the CombinedSearchProvider`
    );
  }
  return context;
}
