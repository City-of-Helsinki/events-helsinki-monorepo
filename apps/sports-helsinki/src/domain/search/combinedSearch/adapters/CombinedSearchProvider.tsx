import useLocale from '@events-helsinki/components/hooks/useLocale';
import { useSearchParams } from 'next/navigation';
import { useRouter } from 'next/router';
import React from 'react';

import { initialCombinedSearchFormValues } from '../constants';
import type { CombinedSearchAdapterInput } from '../types';
import type { CombinedSearchContextType } from './CombinedSearchContext';
import { CombinedSearchContext } from './CombinedSearchContext';
import CombinedSearchFormAdapter from './CombinedSearchFormAdapter';

type UseGetCombinedSearchContextReturnType = {
  combinedSearchFormAdapter: CombinedSearchFormAdapter;
  contextValue: CombinedSearchContextType;
};

function useGetCombinedSearchContext(): UseGetCombinedSearchContextReturnType {
  const searchParams = useSearchParams();
  const router = useRouter();
  const locale = useLocale();

  const combinedSearchFormAdapter = new CombinedSearchFormAdapter(
    router,
    locale,
    searchParams
  );

  const getFormValues = () => combinedSearchFormAdapter.getFormValues();
  const getSearchVariables = () =>
    combinedSearchFormAdapter.getSearchVariables();
  const setFormValues = (values: Partial<CombinedSearchAdapterInput>) =>
    combinedSearchFormAdapter.setFormValues(values);
  const setFormValue = (
    field: keyof CombinedSearchAdapterInput,
    value: CombinedSearchAdapterInput[keyof CombinedSearchAdapterInput]
  ) =>
    setFormValues({
      [field]: value,
    });
  const updateRouteToSearchPage = () => {
    combinedSearchFormAdapter.routerPush();
  };
  const resetFormValues = () => setFormValues(initialCombinedSearchFormValues);

  return {
    combinedSearchFormAdapter,
    contextValue: {
      formValues: getFormValues(),
      searchVariables: getSearchVariables(),
      setFormValues,
      setFormValue,
      resetFormValues,
      updateRouteToSearchPage,
    },
  };
}

export function CombinedSearchProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const { contextValue } = useGetCombinedSearchContext();
  return (
    <CombinedSearchContext.Provider value={contextValue}>
      {children}
    </CombinedSearchContext.Provider>
  );
}
