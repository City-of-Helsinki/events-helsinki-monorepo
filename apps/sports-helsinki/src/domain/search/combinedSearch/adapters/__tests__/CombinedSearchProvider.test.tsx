import mockRouter from 'next-router-mock';
import qs from 'query-string';
import { render } from '@/test-utils';
import { useCombinedSearchContext } from '../CombinedSearchContext';
import { CombinedSearchProvider } from '../CombinedSearchProvider';

const TestingComponent = () => {
  const { formValues, searchVariables } = useCombinedSearchContext();
  return (
    <>
      <p data-test-id="formValues">{JSON.stringify(formValues)}</p>
      <p data-test-id="event-searchVariables">
        {JSON.stringify(searchVariables.event)}
      </p>
      <p data-test-id="course-searchVariables">
        {JSON.stringify(searchVariables.course)}
      </p>
      <p data-test-id="venue-searchVariables">
        {JSON.stringify(searchVariables.course)}
      </p>
    </>
  );
};

describe('CombinedSearchProvider', () => {
  it('reads the context properly', () => {
    const searchParams = new URLSearchParams('text=test%20text');
    const { container } = render(
      <CombinedSearchProvider searchParams={searchParams}>
        <TestingComponent />
      </CombinedSearchProvider>
    );
    expect(container).toMatchSnapshot();
  });

  // FIXME: Skipped while this feature is still planned
  it.skip('pushes the router when search params changes', () => {
    mockRouter.setCurrentUrl(`/`);
    expect(qs.stringify(mockRouter.query)).toBe('');
    const searchParams = new URLSearchParams('text=test%20text');
    render(
      <CombinedSearchProvider searchParams={searchParams}>
        <TestingComponent />
      </CombinedSearchProvider>
    );
    expect(qs.stringify(mockRouter.query)).toBe('text=test%20text');
  });
});
