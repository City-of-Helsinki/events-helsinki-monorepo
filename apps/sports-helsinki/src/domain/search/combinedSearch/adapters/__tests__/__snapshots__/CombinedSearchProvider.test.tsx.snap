// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CombinedSearchProvider reads the context properly 1`] = `
<div>
  <p
    data-test-id="formValues"
  >
    {"text":"test text","sportsCategories":[],"targetGroups":[],"keywords":[]}
  </p>
  <p
    data-test-id="event-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"General","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
  <p
    data-test-id="course-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
  <p
    data-test-id="venue-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
</div>
`;

exports[`CombinedSearchProvider reads the context properly 2`] = `
<div>
  <p
    data-test-id="formValues"
  >
    {"text":"test text","sportsCategories":["gym","playgrounds"],"targetGroups":[],"keywords":[]}
  </p>
  <p
    data-test-id="event-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":["yso:p8504","yso:p8105"],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"General","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
  <p
    data-test-id="course-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":["yso:p8504","yso:p8105"],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
  <p
    data-test-id="venue-searchVariables"
  >
    {"allOngoingAnd":["test text"],"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":["yso:p8504","yso:p8105"],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":null,"pageSize":10}
  </p>
</div>
`;

exports[`CombinedSearchProvider reads the context properly 3`] = `
<div>
  <p
    data-test-id="formValues"
  >
    {"text":"","sportsCategories":[],"targetGroups":[],"organization":"testorg","keywords":[]}
  </p>
  <p
    data-test-id="event-searchVariables"
  >
    {"allOngoingAnd":[],"allOngoing":true,"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"General","superEventType":["umbrella","none"],"publisher":"testorg","pageSize":10}
  </p>
  <p
    data-test-id="course-searchVariables"
  >
    {"allOngoingAnd":[],"allOngoing":true,"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":"testorg","pageSize":10}
  </p>
  <p
    data-test-id="venue-searchVariables"
  >
    {"allOngoingAnd":[],"allOngoing":true,"start":"now","end":"","include":["keywords","location"],"keywordAnd":[],"keywordNot":[],"keywordOrSet1":["yso:p916","kulke:710","yso:p17018","yso:p1963","yso:p9824","yso:p965","yso:p6409","yso:p8781","yso:p26619","yso:p13035","yso:p2041"],"keywordOrSet2":[],"keywordOrSet3":[],"location":[],"sort":"end_time","eventType":"Course","superEventType":["umbrella","none"],"publisher":"testorg","pageSize":10}
  </p>
</div>
`;
